git/odb/file_storer.go

```diff
@@ -0,0 +1,65 @@
+package odb
+
+import (
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+)
+
+// FileStorer implements the Storer interface by writing to the .git/objects
+// directory on disc.
+type FileStorer struct {
+	// root is the top level /objects directory's path on disc.
+	root string
+}
+
+var _ Storer = (*FileStorer)(nil)
+
+// NewFileStorer returns a new FileStorer instance with the given root.
+func NewFileStorer(root string) *FileStorer {
+	return &FileStorer{
+		root: root,
+	}
+}
+
+// Open implements the Storer.Open function, and returns a io.ReadWriteCloser
+// for the given SHA. If the file does not exist, or if there was any other
+// error in opening the file, an error will be returned.
+//
+// It is the caller's responsibility to close the given file "f" after its use
+// is complete.
+func (fs *FileStorer) Open(sha []byte) (f io.ReadWriteCloser, err error) {
+	return fs.open(fs.path(sha), os.O_RDONLY)
+}
+
+// Create implements the Storere.Create function and returns an
+// io.ReadWriteCloser for the given SHA, provided that an object of that SHA is
+// not already indexed in the database.
+//
+// If the file already exists, could not be created, or opened, an error will be
+// returned.
+//
+// It is the caller's responsibility to close the given file "f" after its use
+// is complete.
+func (fs *FileStorer) Create(sha []byte) (f io.ReadWriteCloser, err error) {
+	path := fs.path(sha)
+	dir := filepath.Dir(path)
+
+	if err = os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+	return fs.open(path, os.O_RDWR|os.O_CREATE|os.O_EXCL)
+}
```

Comment:
You should avoid opening the object path directly. While you're writing, simultaneous readers will see a corrupted object. And if the program is interrupted for any reason, you'll leave a half-written object on disk.  The usual git.git way is to open a temporary file in the `objects` directory, and then `rename()` or `link()` the file into place. Those syscalls are atomic with respect to other processes trying to read the object; they either see the whole thing or nothing at all.

As a bonus, this makes it easy to stream files into the object database; you open the tempfile, stream the contents into place, doing zlib compression and computing the sha1 on the fly, and then rename the file based on the final sha1 you computed.